File: comprehensive_stock_analysis/src/stock_analysis/web/routes/alerts.py
"""Alert log and settings endpoints."""

from typing import List

from fastapi import APIRouter, Query

from ..schemas import AlertItem, AlertSettingsRequest, AlertSettingsResponse

router = APIRouter(prefix="/api", tags=["alerts"])


def _log_entry_to_item(entry: dict) -> AlertItem:
    return AlertItem(
        symbol=entry.get("symbol", ""),
        timestamp=entry.get("fired_at", entry.get("timestamp", "")),
        reason=entry.get("reason", ""),
        old_rec=entry.get("old_recommendation") or entry.get("old_rec") or None,
        new_rec=entry.get("new_recommendation") or entry.get("new_rec") or None,
        old_confidence=entry.get("old_confidence"),
        new_confidence=entry.get("new_confidence"),
    )


@router.get("/alerts", response_model=List[AlertItem])
def list_alerts(limit: int = Query(200, ge=1, le=1000)) -> List[AlertItem]:
    from ..alerts import get_alert_log
    return [_log_entry_to_item(e) for e in get_alert_log(limit=limit)]


@router.post("/settings/alerts", response_model=dict)
def save_alert_settings(body: AlertSettingsRequest) -> dict:
    from ..alerts import save_alert_settings as _save

    values = {
        "alert_email": body.alert_email,
        "alert_smtp_host": body.alert_smtp_host,
        "alert_smtp_port": None if body.alert_smtp_port is None else str(body.alert_smtp_port),
        "alert_smtp_user": body.alert_smtp_user,
        "alert_smtp_password": body.alert_smtp_password,
        "alert_webhook_url": body.alert_webhook_url,
    }
    _save(values)
    return {"status": "ok"}


@router.get("/settings/alerts", response_model=AlertSettingsResponse)
def get_alert_settings_route() -> AlertSettingsResponse:
    from ..alerts import get_alert_settings

    cfg = get_alert_settings()
    return AlertSettingsResponse(
        alert_email=cfg["alert_email"],
        alert_smtp_host=cfg["alert_smtp_host"],
        alert_smtp_port=cfg["alert_smtp_port"],
        alert_smtp_user=cfg["alert_smtp_user"],
        alert_smtp_password_set=bool(cfg["alert_smtp_password"]),
        alert_webhook_url=cfg["alert_webhook_url"],
    )

File: comprehensive_stock_analysis/src/stock_analysis/web/static/js/alerts.js
// Alerts view: notification settings form + recent alert log.
import { $, el, fetchJSON } from "./util.js";

export async function loadAlerts() {
  $("#alert-settings-form").onsubmit = onSaveSettings;
  await Promise.all([loadSettings(), loadLog()]);
}

async function loadSettings() {
  let cfg;
  try {
    cfg = await fetchJSON("/api/settings/alerts");
  } catch (_) { return; }
  $("#alert-email").value = cfg.alert_email || "";
  $("#alert-smtp-host").value = cfg.alert_smtp_host || "";
  $("#alert-smtp-port").value = cfg.alert_smtp_port || "";
  $("#alert-smtp-user").value = cfg.alert_smtp_user || "";
  $("#alert-smtp-password").placeholder = cfg.alert_smtp_password_set
    ? "•••••••• (set — leave blank to keep)" : "Not set";
  $("#alert-webhook").value = cfg.alert_webhook_url || "";
}

async function onSaveSettings(ev) {
  ev.preventDefault();
  const status = $("#alert-settings-status");
  status.textContent = "Saving…";
  const payload = {
    alert_email: $("#alert-email").value.trim() || null,
    alert_smtp_host: $("#alert-smtp-host").value.trim() || null,
    alert_smtp_port: $("#alert-smtp-port").value ? Number($("#alert-smtp-port").value) : null,
    alert_smtp_user: $("#alert-smtp-user").value.trim() || null,
    // Blank password field means "don't change" — omit it entirely.
    alert_smtp_password: $("#alert-smtp-password").value || null,
    alert_webhook_url: $("#alert-webhook").value.trim() || null,
  };
  try {
    await fetchJSON("/api/settings/alerts", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    status.textContent = "Saved.";
    $("#alert-smtp-password").value = "";
    await loadSettings();
  } catch (err) {
    status.textContent = "Error: " + err.message;
  }
  setTimeout(() => { if (status.textContent !== "Saving…") status.textContent = ""; }, 3000);
}

async function loadLog() {
  const box = $("#alerts-log");
  const empty = $("#alerts-empty");
  box.innerHTML = "";
  let items = [];
  try {
    items = await fetchJSON("/api/alerts?limit=100");
  } catch (_) { items = []; }
  empty.classList.toggle("hidden", items.length > 0);
  items.forEach((it) => box.append(logItem(it)));
}

function logItem(it) {
  return el("div", { class: "alert-log-item" },
    el("span", { class: "alert-log-dot" }),
    el("div", { class: "alert-log-body" },
      el("div", { class: "alert-log-reason" }, `${it.symbol}: ${it.reason}`),
      el("div", { class: "alert-log-meta" }, (it.timestamp || "").replace("T", " "))));
}
